contracts/test/umbra-hook-test.js

```diff
@@ -0,0 +1,407 @@
+  const relayerTokenFee = toWei('2', 'ether');
+
+  // The ethers wallet that will be used when testing meta tx withdrawals
+  const metaWallet = ethers.Wallet.createRandom();
+
+  // helper function to mint test tokens
+  const mintToken = async (toAddr, amount) => {
+    await ctx.token.mint(toAddr, toWei(amount));
+  };
+
+  // helper function to send test tokens to stealth address
+  const sendToken = async (fromAddr, stealthAddr, amount) => {
+    const toll = await ctx.umbra.toll();
+    const weiAmount = toWei(amount);
+
+    await ctx.token.approve(ctx.umbra.address, weiAmount, { from: fromAddr });
+
+    await ctx.umbra.sendToken(stealthAddr, ctx.token.address, weiAmount, ...argumentBytes, {
+      from: fromAddr,
+      value: toll,
+    });
+  };
+
+  // helper function to mint and send tokens to steatlh address
+  const mintAndSendToken = async (fromAddr, stealthAddr, amount) => {
+    await mintToken(fromAddr, amount);
+    await sendToken(fromAddr, stealthAddr, amount);
+  };
+
+  before(async () => {
+    [owner, sender, receiver, acceptor, relayer] = await web3.eth.getAccounts();
+    ctx.chainId = await web3.eth.getChainId();
+  });
+
+  beforeEach(async () => {
+    ctx.umbra = await Umbra.new(toll, owner, owner, { from: owner });
+    ctx.token = await TestToken.new('TestToken', 'TT');
+    ctx.hook = await MockHook.new();
+    ctx.version = await ctx.umbra.version();
+  });
+
+  it('should see the deployed Umbra contract', async () => {
+    expect(ctx.umbra.address.startsWith('0x')).to.be.true;
+    expect(ctx.umbra.address.length).to.equal(42);
+    expect(ctx.token.address.startsWith('0x')).to.be.true;
+    expect(ctx.token.address.length).to.equal(42);
+    expect(ctx.hook.address.startsWith('0x')).to.be.true;
+    expect(ctx.hook.address.length).to.equal(42);
+  });
+
+  it('should call the hook contract when the stealth receiver withdraws directly', async () => {
+    await mintAndSendToken(sender, receiver, '100');
+
+    await ctx.umbra.withdrawTokenAndCall(acceptor, ctx.token.address, ctx.hook.address, '0xbeefc0ffee', {
+      from: receiver,
+    });
+
+    const callData = await ctx.hook.lastCallData();
+
+    expect(fromWei(callData.amount)).to.equal('100');
+    expect(callData.stealthAddr).to.equal(receiver);
+    expect(callData.acceptor).to.equal(acceptor);
+    expect(callData.tokenAddr).to.equal(ctx.token.address);
+    expect(callData.sponsor).to.equal(AddressZero);
+    expect(fromWei(callData.sponsorFee)).to.equal('0');
+    expect(callData.data).to.equal('0xbeefc0ffee');
+
+    const balance = await ctx.token.balanceOf(acceptor);
+    expect(fromWei(balance)).to.equal('100');
+  });
+
+  it('should call the hook contract when the stealth receiver withdraws directly with empty data', async () => {
+    await mintAndSendToken(sender, receiver, '100');
+
+    await ctx.umbra.withdrawTokenAndCall(acceptor, ctx.token.address, ctx.hook.address, [], {
```

Comment:
Does it matter if this is passed is as an empty array vs empty string? I'd have thought `""` would be the way to do this, so curious if both work to pass in empty bytes. Surprised it's returned as `null` and not an empty string, maybe that's a web3 convention? Doesn't matter much, just curious

Reply:
yeah, I wasted time figuring this one out. Passing `""` to ethers' `defaultAbiCoder.encode` function results in the following error AND to web3.js as a contract call argument both result in the following error:

```
Error: invalid arrayify value (argument="value", value="", code=INVALID_ARGUMENT, version=bytes/5.0.10)
```

Given that they're the same, I assume some underlying dependency requires this. I don't like it and wish it would use `""` or just `null` but this is what it's looking for.

Reply:
Ahhh, I think it's because bytes are usually represented in byte arrays where `0x1234` is represented as two bytes of `[18, 52]`. See:

- https://docs.ethers.io/v5/single-page/#/v5/api/utils/bytes/-%23-utils-arrayify
- https://docs.ethers.io/v5/single-page/#/v5/api/signer/-%23-Signer--signing-methods

Reply:
It seems passing in `0x` also works

Reply:
Updated with ecec84a6d4a55c2a17e07cf2403260c38947b232